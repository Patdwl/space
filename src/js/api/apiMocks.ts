// @ts-nocheck

import { SatObject, SensorObject } from './keepTrackTypes';
import { SpaceObjectType } from './SpaceObjectType';
declare const jest: any;

export const defaultSat: SatObject = {
  id: 1,
  active: true,
  sccNum: '00005',
  intlDes: '1998-AEF',
  country: 'ISS',
  launchSite: 'TTMTR',
  launchVehicle: 'Proton-K',
  name: 'ISS (ZARYA)',
  type: SpaceObjectType.PAYLOAD,
  TLE1: '1 25544U 98067A   21203.40407588  .00003453  00000-0  71172-4 0  9991',
  TLE2: '2 25544  51.6423 168.5744 0001475 184.3976 313.3642 15.48839820294053',
  rcs: '99.0524',
  owner: 'National Aeronautics and Space Administration (NASA)/Multinational',
  user: 'Government',
  purpose: 'Space Science',
  launchMass: '',
  dryMass: '',
  power: '',
  lifetime: 30,
  manufacturer: 'Boeing Satellite Systems (prime)/Multinational',
  mission: 'Final size of a Boeing 747; first component.',
  inclination: 51.6423,
  lon: 168.5744,
  perigee: 1475,
  apogee: 184.3976,
  period: 313.3642,
  meanMotion: 15.48839820294053,
  semimajorAxis: 21203.40407588,
  eccentricity: 0.00003453,
  raan: 0,
  argPe: 0,
  inView: 1,
  velocity: {
    total: 0,
    x: 0,
    y: 0,
    z: 0,
  },
  getTEARR: () => ({
    lon: 168.5744,
    lat: 51.6423,
    alt: 1475,
  }),
  getAltitude: () => 0,
  isInSun: () => true,
  getDirection: () => 'N',
  position: {
    x: 0,
    y: 0,
    z: 0,
  },
  static: false,
  staticNum: null,
};

export const defaultSensor: SensorObject = {
  alt: 0.060966,
  beamwidth: 2,
  changeObjectInterval: 1000,
  country: 'United States',
  lat: 41.754785,
  linkAehf: true,
  linkWgs: true,
  lon: -70.539151,
  name: 'Cape Cod AFS, Massachusetts',
  observerGd: { lat: 0.7287584767123405, lon: -1.2311404365114507, alt: 0.060966 },
  obsmaxaz: 227,
  obsmaxel: 85,
  obsmaxrange: 5556,
  obsminaz: 347,
  obsminel: 3,
  obsminrange: 200,
  shortName: 'COD',
  staticNum: 0,
  sun: 'No Impact',
  type: SpaceObjectType.PHASED_ARRAY_RADAR,
  url: 'http://www.radartutorial.eu/19.kartei/01.oth/karte004.en.html',
  volume: false,
  zoom: 'leo',
};

export const useMockWorkers = (): void => {
  class Worker {
    url: any;
    // eslint-disable-next-line no-unused-vars
    onmessage: (msg: any) => { data: {} };
    constructor(stringUrl: any) {
      this.url = stringUrl;
      this.onmessage = (msg: any) => {
        console.debug(msg);
        return {
          data: {},
        };
      };
    }

    postMessage(msg: any) {
      if (msg.dat) {
        this.onmessage({
          data: {
            extraData: JSON.stringify([defaultSat, defaultSensor]),
            satPos: [0, 0, 0, 0, 0, 0],
            satVel: [0, 0, 0, 0, 0, 0],
          },
        });
      }
      this.onmessage({
        data: {
          satPos: [0, 0, 0, 0, 0, 0],
          satVel: [0, 0, 0, 0, 0, 0],
        },
      });
    }
  }

  (<any>window).Worker = Worker;
};

export const keepTrackApiStubs = {
  programs: {
    adviceManager: {
      adviceList: {
        satViewDisabled: jest.fn(),
        editSatDisabled: jest.fn(),
        survFenceDisabled: jest.fn(),
        bubbleDisabled: jest.fn(),
        satFovDisabled: jest.fn(),
        planetariumDisabled: jest.fn(),
        mapDisabled: jest.fn(),
        breakupDisabled: jest.fn(),
        ssnLookanglesDisabled: jest.fn(),
        sensorInfoDisabled: jest.fn(),
        lookanglesDisabled: jest.fn(),
        satelliteSelected: jest.fn(),
        sensor: jest.fn(),
        cspocSensors: jest.fn(),
        mwSensors: jest.fn(),
      },
      adviceCount: {
        socrates: 0,
      },
      showAdvice: jest.fn(),
      adviceArray: <any[]>[],
    },
    mainCamera: {
      camMatrix: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
      cameraType: {
        current: 1,
        Default: 0,
        Offset: 1,
        FixedToSat: 2,
        Fps: 3,
        Planetarium: 4,
        Satellite: 5,
        Astronomy: 6,
      },
      camSnap: jest.fn(),
      changeCameraType: jest.fn(),
      changeZoom: jest.fn(),
      lat2pitch: jest.fn(),
      lon2yaw: jest.fn(),
      calculate: jest.fn(),
      update: jest.fn(),
      getCamDist: jest.fn(),
      localRotateDif: {
        pitch: 0,
        yaw: 0,
      },
      zoomLevel: () => 0,
      zoomTarget: () => 0,
      autoRotate: jest.fn(),
      fts2default: jest.fn(),
      panCurrent: {
        x: 0,
        y: 0,
        z: 0,
      },
      getForwardVector: jest.fn(),
    },
    colorSchemeManager: {
      objectTypeFlags: {
        payload: true,
        rocketBody: true,
        debris: true,
        inFOV: true,
      },
      resetObjectTypeFlags: jest.fn(),
      reloadColors: jest.fn(),
      default: () => [0, 0, 0, 0],
      currentColorScheme: () => [0, 0, 0, 0],
      group: null,
    },
    drawManager: {
      pMatrix: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
      resizeCanvas: jest.fn(),
      glInit: jest.fn(),
      gl: global.mocks.glMock,
      canvas: {
        style: {
          cursor: '',
        },
        width: 0,
        height: 0,
      },
      selectSatManager: {
        selectSat: jest.fn(),
      },
      postProcessingManager: {
        init: jest.fn(),
        curBuffer: null,
        programs: {
          occlusion: {
            attrSetup: jest.fn(),
            attrOff: jest.fn(),
            uniformSetup: jest.fn(),
          },
        },
      },
      sceneManager: {
        sun: {
          initGodrays: jest.fn(),
          pos: [0, 0, 0],
          sunvar: {
            gmst: 0,
          },
          draw: jest.fn(),
          drawGodrays: jest.fn(),
          godrays: {
            frameBuffer: {},
          },
          now: new Date(2022, 1, 1),
        },
        earth: {
          init: jest.fn(),
          loadHiRes: jest.fn(),
          loadHiResNight: jest.fn(),
          lightDirection: [0, 0, 0],
          drawOcclusion: () => {}, // NOSONAR
        },
        moon: {
          draw: jest.fn(),
        },
      },
      startWithOrbits: jest.fn(),
    },
    dotsManager: {
      inSunData: new Float32Array([0, 0, 0]),
      inViewData: new Float32Array([0, 0, 0]),
      init: jest.fn(),
      starIndex1: 0,
      pickReadPixelBuffer: [0, 0, 0],
      pickingDotSize: '1.0',
      gl: null,
      emptyMat4: null,
      positionBuffer: null,
      sizeBuffer: null,
      loaded: true,
      pMvCamMatrix: null,
      drawProgram: null,
      positionBufferOneTime: null,
      positionData: new Float32Array([0, 0, 0]),
      pickingProgram: {
        aPos: null,
      },
      pickingColorBuffer: null,
      drawShaderCode: { frag: null, vert: null },
      pickingShaderCode: { vert: null, frag: null },
      pickingTexture: null,
      pickingRenderBuffer: null,
      velocityData: new Float32Array([0, 0, 0]),
      drawDivisor: null,
      satDataLenInDraw: null,
      satDataLenInDraw3: null,
      orbitalSats3: null,
      drawI: null,
      draw: jest.fn(),
      drawGpuPickingFrameBuffer: jest.fn(),
      updatePositionBuffer: jest.fn(),
      updateSizeBuffer: jest.fn(),
      updatePMvCamMatrix: jest.fn(),
      sizeBufferOneTime: null,
      sizeData: null,
      starIndex2: null,
      pickingColorData: null,
      pickingFrameBuffer: null,
      createPickingProgram: jest.fn(),
      sensorMarkerArray: new Float32Array([0, 0, 0]),
    },
    groups: null,
    soundManager: {
      play: jest.fn(),
    },
    groupsManager: {
      Canada: null,
      Japan: null,
      France: null,
      China: null,
      India: null,
      Israel: null,
      Russia: null,
      UnitedKingdom: null,
      UnitedStates: null,
      SpaceStations: {
        group: {},
        sats: [
          { satId: 1, sccNum: '25544' },
          { satId: 1, sccNum: '25544' },
        ],
        updateOrbits: jest.fn(),
      },
      clearSelect: jest.fn(),
      createGroup: () => ({
        sats: [
          { satId: 1, sccNum: '25544' },
          { satId: 1, sccNum: '25544' },
        ],
        updateOrbits: jest.fn(),
      }),
      selectGroup: jest.fn(),
      selectedGroup: [1, 2, 3],
    },
    lineManager: {
      draw: jest.fn(),
      create: jest.fn(),
      clear: jest.fn(),
      removeStars: jest.fn(),
      getLineListLen: jest.fn(),
      drawWhenSelected: jest.fn(),
    },
    missileManager: {
      getMissileTEARR: jest.fn(),
    },
    objectManager: {
      isLaunchSiteManagerLoaded: true,
      lastSelectedSat: jest.fn(),
      extractCountry: jest.fn(),
      setHoveringSat: jest.fn(),
      setLasthoveringSat: jest.fn(),
      extractLaunchSite: () => ({ site: 'test', sitec: 'test' }),
      extractLiftVehicle: jest.fn(),
      launchSiteManager: {
        launchSiteList: ['AFWTR'],
      },
      selectedSat: -1,
      setSelectedSat: jest.fn(),
      staticSet: {
        id: 0,
      },
      analSatSet: [defaultSat],
      radarDataSet: {
        id: 0,
      },
      missileSet: {
        id: 0,
      },
      fieldOfViewSet: {
        id: 0,
      },
      satLinkManager: {
        idToSatnum: jest.fn(),
        showLinks: jest.fn(),
        aehf: [25544],
        dscs: [25544],
        wgs: [25544],
        starlink: [25544],
        sbirs: [25544],
      },
    },
    meshManager: {
      draw: jest.fn(),
    },
    orbitManager: {
      orbitWorker: {
        postMessage: jest.fn(),
      },
      setSelectOrbit: jest.fn(),
      draw: jest.fn(),
      setHoverOrbit: jest.fn(),
      clearHoverOrbit: jest.fn(),
      clearSelectOrbit: jest.fn(),
      updateOrbitBuffer: jest.fn(),
      clearInViewOrbit: jest.fn(),
      removeInViewOrbit: jest.fn(),
      historyOfSatellitesPlay: jest.fn(),
      shader: {
        uColor: 1,
      },
    },
    satellite: {
      currentTEARR: {
        az: 0,
        el: 0,
        rng: 0,
      },
      sgp4: () => [0, 0, 0],
      setobs: jest.fn(),
      distance: jest.fn(),
      setTEARR: jest.fn(),
      eci2ll: () => ({ lat: 0, lon: 0 }),
      altitudeCheck: () => 10000,
      calculateSensorPos: jest.fn(),
      degreesLong: (num: number) => num,
      degreesLat: (num: number) => num,
      currentEpoch: () => ['21', '150'],
      getOrbitByLatLon: () => ['1 25544U 98067A   21203.40407588  .00003453  00000-0  71172-4 0  9991', '2 25544  51.6423 168.5744 0001475 184.3976 313.3642 15.48839820294053'],
      nextpassList: () => [new Date().getTime(), new Date().getTime() + 1],
      twoline2satrec: () => ({ jdsatepoch: 2458000.5, ecco: 1, nodeo: 1, argpo: 1, meana: 1, argPe: 1, argPeA: 1 }),
      getRae: () => [0, 0, 0],
      checkIsInView: () => true,
      getlookangles: jest.fn(),
      getlookanglesMultiSite: jest.fn(),
      findCloseObjects: () => '',
      createTle: () => ({ TLE1: '', TLE2: '' }),
      eci2Rae: jest.fn(),
      findBestPasses: () => jest.fn(),
      findNearbyObjectsByOrbit: () => jest.fn(),
      nextpass: () => '2021-09-21 20:00:00Z',
      getDops: () => ({
        HDOP: 0,
        VDOP: 0,
        TDOP: 0,
        GDOP: 0,
        PDOP: 0,
      }),
      updateDopsTable: jest.fn(),
      map: () => ({ lat: 0, lon: 0 }),
      sat2ric: () => ({ position: [0, 0, 0], velocity: [0, 0, 0] }),
      getTEARR: () => ({
        az: 0,
        el: 0,
        rng: 0,
        inView: true,
      }),
    },
    mapManager: {
      isMapMenuOpen: false,
    },
    controlSiteManager: {},
    satSet: {
      satCruncher: {
        postMessage: jest.fn(),
        onmessage: jest.fn(),
      },
      getIdFromEci: jest.fn(),
      selectSat: jest.fn(),
      satData: [
        { ...defaultSat, ...{ id: 0 } },
        defaultSat,
        { id: 2, type: SpaceObjectType.STAR, name: 'test' },
        { id: 3, type: SpaceObjectType.PHASED_ARRAY_RADAR, static: true, name: 'test' },
      ],
      convertIdArrayToSatnumArray: jest.fn(),
      getIdFromObjNum: () => 0,
      getSensorFromSensorName: () => 0,
      getSatExtraOnly: () => defaultSat,
      getSatFromObjNum: () => defaultSat,
      getSat: () => defaultSat,
      getIdFromStarName: jest.fn(),
      initGsData: jest.fn(),
      setColorScheme: jest.fn(),
      getSatPosOnly: () => defaultSat,
      getSatInView: () => jest.fn(),
      getScreenCoords: () => ({
        error: false,
      }),
      getSatVel: () => jest.fn(),
      getSatInSun: () => jest.fn(),
      sunECI: { x: 0, y: 0, z: 0 },
      getSatInViewOnly: () => defaultSat,
      default: '',
      setHover: jest.fn(),
      insertNewAnalystSatellite: () => ({ sccNum: 25544 }),
      getIdFromIntlDes: jest.fn(),
      queryStr: 'search=25544&intldes=1998-A&sat=25544&misl=0,0,0&date=1234567&rate=1&hires=true',
      cosparIndex: { '1998-AB': 5 },
      numSats: 1,
      missileSats: 10000,
      search: {
        year: () => [defaultSat],
        yearOrLess: () => [defaultSat],
        name: () => [defaultSat],
        country: () => [defaultSat],
        shape: () => [defaultSat],
        bus: () => [defaultSat],
        type: () => [defaultSat],
      },
    },
    searchBox: {
      hideResults: jest.fn(),
      fillResultBox: jest.fn(),
      doArraySearch: jest.fn(),
      doSearch: jest.fn(),
      getLastResultGroup: jest.fn(),
      isHovering: jest.fn(),
      setHoverSat: jest.fn(),
      isResultBoxOpen: jest.fn(),
    },
    sensorFov: {
      enableFovView: jest.fn(),
    },
    sensorManager: {
      checkSensorSelected: () => false,
      currentSensor: [
        {
          alt: 0,
          lat: 0,
          lon: 0,
          shortName: 'COD',
          obsmaxaz: 0,
          obsmaxel: 0,
          obsmaxrange: 0,
          obsminaz: 0,
          obsminel: 0,
          obsminrange: 0,
        },
      ],
      sensorListUS: [defaultSensor, defaultSensor],
      setCurrentSensor: jest.fn(),
      setSensor: jest.fn(),
      sensorList: { COD: defaultSensor },
    },
    starManager: {
      init: jest.fn(),
      findStarsConstellation: jest.fn(),
      drawConstellations: jest.fn(),
    },
    timeManager: {
      propOffset: 0,
      getPropOffset: () => 0,
      propTimeVar: new Date(2020, 0, 1),
      propRate: 0,
      selectedDate: 0,
      getDayOfYear: () => 0,
      updatePropTime: jest.fn(),
      jday: () => 0,
      dateFromDay: () => new Date(2020, 0, 1),
      setNow: jest.fn(),
      changePropRate: jest.fn(),
      changeStaticOffset: jest.fn(),
      calculateSimulationTime: () => new Date(2020, 0, 1),
      getOffsetTimeObj: () => new Date(2020, 0, 1),
      synchronize: jest.fn(),
    },
    uiManager: {
      hideSideMenus: jest.fn(),
      footerToggle: jest.fn(),
      toast: jest.fn(),
      colorSchemeChangeAlert: jest.fn(),
      bottomIconPress: jest.fn(),
      legendColorsChange: jest.fn(),
      doSearch: jest.fn(),
      loadStr: jest.fn(),
      searchToggle: jest.fn(),
      legendMenuChange: jest.fn(),
      clearRMBSubMenu: jest.fn(),
      updateURL: jest.fn(),
      keyHandler: jest.fn(),
      getsensorinfo: jest.fn(),
      hideLoadingScreen: jest.fn(),
      legendHoverMenuClick: jest.fn(),
      mobileManager: {
        fullscreenToggle: jest.fn(),
        checkMobileMode: jest.fn(),
      },
      uiInput: {
        getSatIdFromCoord: () => 0,
        mouseSat: 0,
      },
    },
    watchlist: {
      updateWatchlist: jest.fn(),
    },
  },
};

keepTrackApiStubs.programs.uiManager.searchBox = keepTrackApiStubs.programs.searchBox;

keepTrackApiStubs.programs.groupsManager.Canada = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.Japan = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.France = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.China = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.India = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.Israel = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.Russia = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.UnitedKingdom = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.UnitedStates = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groups = keepTrackApiStubs.programs.groupsManager;
