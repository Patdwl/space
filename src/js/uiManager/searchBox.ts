/* */

import { keepTrackApi } from '@app/js/api/keepTrackApi';
import $ from 'jquery';
import { CatalogManager, SearchBox } from '../api/keepTrackTypes';
import { SpaceObjectType } from '../api/SpaceObjectType';
import { SatGroup } from '../groupsManager/sat-group';

let hoverSatId = -1;
let hovering = false;
let resultsOpen = false;
let lastResultGroup: SatGroup;

interface SearchResult {
  satId: any;
  isON: any;
  strIndex: number;
  patlen: any;
  isSccNum: any;
  isIntlDes: any;
}

export const isResultBoxOpen = (): boolean => resultsOpen;
export const getLastResultGroup = () => lastResultGroup;
export const getHoverSat = () => hoverSatId;
export const isHovering = (val?: boolean): boolean => (hovering = typeof val !== 'undefined' ? val : hovering);
export const setHoverSat = (satId: number): number => (hoverSatId = satId);
export const getCurrentSearch = () => (resultsOpen ? $('#search').val() : null);

export const hideResults = () => {
  try {
    const { colorSchemeManager, groupsManager, dotsManager, satSet } = keepTrackApi.programs;
    $('#search-results').slideUp();
    groupsManager.clearSelect();
    resultsOpen = false;

    settingsManager.lastSearch = '';
    settingsManager.lastSearchResults = [];
    dotsManager.updateSizeBuffer(satSet.satData.length);

    if (settingsManager.currentColorScheme === colorSchemeManager.group) {
      satSet.setColorScheme(colorSchemeManager.default, true);
    } else {
      satSet.setColorScheme(settingsManager.currentColorScheme, true);
    }
  } catch (error) {
    console.warn(error);
  }
};

export const doArraySearch = (array: string[]) => array.reduce((searchStr, i) => `${searchStr}${keepTrackApi.programs.satSet.satData[i].sccNum},`, '').slice(0, -1);

export const doSearch = (searchString: string, isPreventDropDown?: boolean): number[] => {
  const { satSet, dotsManager, groupsManager } = keepTrackApi.programs;
  if (satSet.satData.length === 0) throw new Error('No sat data loaded! Check if TLEs are corrupted!');

  if (searchString.length === 0) {
    settingsManager.lastSearch = '';
    settingsManager.lastSearchResults = [];
    dotsManager.updateSizeBuffer(satSet.satData.length);
    $('#search').val('');
    searchBox.hideResults();
    return [];
  }

  $('#search').val(searchString);

  // Don't search for things until at least the minimum characters
  // are typed otherwise there are just too many search results.
  if (searchString.length <= settingsManager.minimumSearchCharacters && searchString !== 'RV_') {
    return [];
  }

  // Uppercase to make this search not case sensitive
  searchString = searchString.toUpperCase();

  // NOTE: We are no longer using spaces because it is difficult
  // to predict when a space is part of satellite name.

  // Split string into array using comma or space as delimiter
  // let searchList = searchString.split(/(?![0-9]+)\s(?=[0-9]+)|,/u);

  // Split string into array using comma
  let searchList = searchString.split(/,/u);

  // Update last search with the most recent search results
  settingsManager.lastSearch = searchList;

  // Initialize search results
  const satData = satSet.satData;
  let results = [];
  searchList.forEach((searchString) => {
    satData.every((sat, i) => {
      if (i > satSet.missileSats) return false;
      const len = searchString.length;
      if (sat.static) return true; // Skip static dots (Maybe these should be searchable?)
      if (sat.marker) return false; // Stop searching once you reach the markers
      if (settingsManager.isSatOverflyModeOn && sat.type !== SpaceObjectType.PAYLOAD) return true; // Skip Debris and Rocket Bodies if In Satelltie FOV Mode
      if (sat.missile && !sat.active) return true; // Skip inactive missiles.
      if (sat.country == 'ANALSAT' && !sat.active) return true; // Skip Fake Analyst satellites
      if (!sat.name) return true; // Everything has a name. If it doesn't then assume it isn't what we are searching for.

      if (sat.name.toUpperCase().indexOf(searchString) !== -1) {
        results.push({
          strIndex: sat.name.indexOf(searchString),
          isON: true,
          patlen: len,
          satId: i,
        });
        return true; // Prevent's duplicate results
      }

      if (!sat.desc) {
        // Do nothing there is no description property
      } else if (sat.desc.toUpperCase().indexOf(searchString) !== -1) {
        results.push({
          strIndex: sat.desc.indexOf(searchString),
          isMissile: true,
          patlen: len,
          satId: i,
        });
        return true; // Prevent's duplicate results
      } else {
        return true; // Last check for missiles
      }

      if (sat.sccNum && sat.sccNum.indexOf(searchString) !== -1) {
        results.push({
          strIndex: sat.sccNum.indexOf(searchString),
          isSccNum: true,
          patlen: len,
          satId: i,
        });
        return true; // Prevent's duplicate results
      }

      if (sat.intlDes && sat.intlDes.indexOf(searchString) !== -1) {
        results.push({
          strIndex: sat.intlDes.indexOf(searchString),
          isIntlDes: true,
          patlen: len,
          satId: i,
        });
        return true; // Prevent's duplicate results
      }

      if (sat.launchVehicle && sat.launchVehicle.toUpperCase().indexOf(searchString) !== -1) {
        results.push({
          strIndex: sat.launchVehicle.indexOf(searchString),
          isLV: true,
          patlen: len,
          satId: i,
        });
        return true; // Prevent's duplicate results
      }

      return true; // No match
    });
  });
  // Remove any results greater than the maximum allowed
  results = results.splice(0, settingsManager.searchLimit);

  // Make a group to hilight results
  const idList = results.map((sat) => sat.satId);
  settingsManager.lastSearchResults = idList;

  dotsManager.updateSizeBuffer(satSet.satData.length);

  const dispGroup = groupsManager.createGroup('idList', idList);
  lastResultGroup = dispGroup;
  groupsManager.selectGroup(dispGroup);

  if (!isPreventDropDown) searchBox.fillResultBox(results, satSet);
  return idList;
};

export const fillResultBox = (results: SearchResult[], satSet: CatalogManager) => {
  let satData = satSet.satData;
  var resultBox = $('#search-results');
  const html = results.reduce((html, result) => {
    const sat = satData[result.satId];
    html += '<div class="search-result" data-sat-id="' + sat.id + '">';
    html += '<div class="truncate-search">';
    if (sat.missile) {
      html += sat.name;
    } else if (result.isON) {
      // If the name matched - highlight it
      html += sat.name.substring(0, result.strIndex);
      html += '<span class="search-hilight">';
      html += sat.name.substring(result.strIndex, result.strIndex + result.patlen);
      html += '</span>';
      html += sat.name.substring(result.strIndex + result.patlen);
    } else {
      // If not, just write the name
      html += sat.name;
    }
    html += '</div>';
    html += '<div class="search-result-scc">';
    if (sat.missile) {
      html += sat.desc;
    } else if (result.isSccNum) {
      // If the object number matched
      result.strIndex = result.strIndex || 0;
      result.patlen = result.patlen || 5;

      html += sat.sccNum.substring(0, result.strIndex);
      html += '<span class="search-hilight">';
      html += sat.sccNum.substring(result.strIndex, result.strIndex + result.patlen);
      html += '</span>';
      html += sat.sccNum.substring(result.strIndex + result.patlen);
    } else if (result.isIntlDes) {
      // If the international designator matched
      result.strIndex = result.strIndex || 0;
      result.patlen = result.patlen || 5;

      html += sat.intlDes.substring(0, result.strIndex);
      html += '<span class="search-hilight">';
      html += sat.intlDes.substring(result.strIndex, result.strIndex + result.patlen);
      html += '</span>';
      html += sat.intlDes.substring(result.strIndex + result.patlen);
    } else {
      // Don't Write the lift vehicle - maybe it should?
      html += sat.sccNum;
    }
    html += '</div></div>';
    return html;
  }, '');
  resultBox[0].innerHTML = html;
  resultBox.slideDown();
  resultsOpen = true;
  satSet.setColorScheme(settingsManager.currentColorScheme, true); // force color recalc
};

export const searchBox: SearchBox = {
  isResultBoxOpen,
  getLastResultGroup,
  getCurrentSearch,
  isHovering,
  setHoverSat,
  getHoverSat,
  hideResults,
  doArraySearch,
  doSearch,
  fillResultBox,
};
