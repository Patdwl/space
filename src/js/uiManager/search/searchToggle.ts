import { getEl } from '@app/js/lib/helpers';
import { hideSideMenus } from '../../plugins/breakup/breakup';
import { searchBox } from './searchBox';

let isSearchOpen = false;
let forceClose = false;
let forceOpen = false;
export const searchToggle = (force?: boolean) => {
  // Reset Force Options
  forceClose = false;
  forceOpen = false;

  // Pass false to force close and true to force open
  if (typeof force != 'undefined') {
    if (!force) forceClose = true;
    if (force) forceOpen = true;
  }

  if ((!isSearchOpen && !forceClose) || forceOpen) {
    isSearchOpen = true;
    getEl('search-holder').classList.remove('search-slide-up');
    getEl('search-holder').classList.add('search-slide-down');
    getEl('search-icon').classList.add('search-icon-search-on');
    getEl('fullscreen-icon').classList.add('top-menu-icons-search-on');
    getEl('tutorial-icon').classList.add('top-menu-icons-search-on');
    getEl('legend-icon').classList.add('top-menu-icons-search-on');
  } else {
    isSearchOpen = false;
    getEl('search-holder').classList.remove('search-slide-down');
    getEl('search-holder').classList.add('search-slide-up');
    getEl('search-icon').classList.remove('search-icon-search-on');
    setTimeout(function () {
      getEl('fullscreen-icon').classList.remove('top-menu-icons-search-on');
      getEl('tutorial-icon').classList.remove('top-menu-icons-search-on');
      getEl('legend-icon').classList.remove('top-menu-icons-search-on');
    }, 500);
    hideSideMenus();
    searchBox.hideResults();
    // getEl('menu-space-stations').classList.remove('bmenu-item-selected');
    // This is getting called too much. Not sure what it was meant to prevent?
    // satSet.setColorScheme(colorSchemeManager.default, true);
    // uiManager.colorSchemeChangeAlert(settingsManager.currentColorScheme);
  }
};
