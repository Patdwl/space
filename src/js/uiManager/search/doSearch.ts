import { keepTrackApi } from '@app/js/api/keepTrackApi';
import { legendMenuChange } from '..';
import { searchBox } from './searchBox';
import { toast } from '../ui/toast';
import { updateURL } from '../url/update-url';

export const doSearch = (searchString: string, isPreventDropDown = false) => {
  if (searchString == '') {
    searchBox.hideResults();
    return;
  }

  const { satSet } = keepTrackApi.programs;

  let idList = searchBox.doSearch(searchString, isPreventDropDown);
  if (idList.length === 0) {
    if (settingsManager.lastSearch?.length > settingsManager.minimumSearchCharacters) {
      toast('No Results Found', 'serious', false);
    }
    searchBox.hideResults();
    return;
  }

  if (settingsManager.isSatOverflyModeOn) {
    satSet.satCruncher.postMessage({
      typ: 'satelliteSelected',
      satelliteSelected: idList,
    });
  }
  // Don't let the search overlap with the legend
  legendMenuChange('clear');
  updateURL();
};
