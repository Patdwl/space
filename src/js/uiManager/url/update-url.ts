import { keepTrackApi } from '@app/js/api/keepTrackApi';
import { searchBox } from '../search/searchBox';

export const updateURL = () => {
  if (settingsManager.isDisableUrlBar) return;
  const { objectManager, satSet, timeManager } = keepTrackApi.programs;

  var arr = window.location.href.split('?');
  var url = arr[0];
  var paramSlices = [];

  if (settingsManager.trusatMode) {
    paramSlices.push('trusat');
  }
  if (objectManager.selectedSat !== -1 && typeof satSet.getSatExtraOnly(objectManager.selectedSat).sccNum != 'undefined') {
    paramSlices.push('sat=' + satSet.getSatExtraOnly(objectManager.selectedSat).sccNum);
  }
  var currentSearch = searchBox.getCurrentSearch();
  if (currentSearch !== '') {
    paramSlices.push('search=' + currentSearch);
  }
  if (timeManager.propRate < 0.99 || timeManager.propRate > 1.01) {
    paramSlices.push('rate=' + timeManager.propRate);
  }

  if (timeManager.staticOffset < -1000 || timeManager.staticOffset > 1000) {
    paramSlices.push('date=' + (timeManager.dynamicOffsetEpoch + timeManager.staticOffset).toString());
  }

  if (paramSlices.length > 0) {
    url += '?' + paramSlices.join('&');
  }

  window.history.replaceState(null, 'Keeptrack', url);
};
