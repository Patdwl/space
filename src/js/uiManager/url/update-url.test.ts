import { keepTrackApiStubs } from '@app/js/api/apiMocks';
import { keepTrackApi } from '@app/js/api/keepTrackApi';
import { KeepTrackPrograms } from '@app/js/api/keepTrackTypes';
import { searchBox } from '@app/js/uiManager/search/searchBox';
import '../../settingsManager/settingsManager';
import { updateURL } from './update-url';

keepTrackApi.programs = <KeepTrackPrograms>(<unknown>{ ...keepTrackApi.programs, ...keepTrackApiStubs.programs });

describe('updateUrl', () => {
  it('should update url', () => {
    updateURL();
    expect(window.location.href).toMatchSnapshot();
  });

  it('should handle trusatMode', () => {
    settingsManager.trusatMode = true;
    updateURL();
    expect(window.location.href).toMatchSnapshot();
    settingsManager.trusatMode = false;
  });

  it('should handle selectedSat', () => {
    keepTrackApi.programs.objectManager.selectedSat = 0;
    updateURL();
    expect(window.location.href).toMatchSnapshot();
    keepTrackApi.programs.objectManager.selectedSat = -1;
  });

  it('should handle a search query', () => {
    jest.spyOn(searchBox, 'getCurrentSearch').mockImplementation(() => 'query');
    updateURL();
    expect(window.location.href).toMatchSnapshot();
  });

  it('should handle a time offset', () => {
    keepTrackApi.programs.timeManager.staticOffset = 1500;
    keepTrackApi.programs.timeManager.dynamicOffsetEpoch = 1500;
    updateURL();
    expect(window.location.href).toMatchSnapshot();
  });
});
