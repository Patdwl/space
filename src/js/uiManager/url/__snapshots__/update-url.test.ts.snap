// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateUrl should handle a search query 1`] = `"http://localhost/?search=query&rate=0"`;

exports[`updateUrl should handle a time offset 1`] = `"http://localhost/?search=query&rate=0&date=3000"`;

exports[`updateUrl should handle selectedSat 1`] = `"http://localhost/?sat=00005&rate=0"`;

exports[`updateUrl should handle trusatMode 1`] = `"http://localhost/?trusat&rate=0"`;

exports[`updateUrl should update url 1`] = `"http://localhost/?rate=0"`;
